File: week9/ch.Choe/Week9/Week9/Waitress.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Week9
{
    class Waitress
    {
        Menu1 menu1;
        Menu2 menu2;


        public Waitress(Menu1 m1, Menu2 m2)
        {
            menu1 = m1;
            menu2 = m2;
        }
        public void PrintMenu()
        {
            System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
            sw.Start();

            UserIterator intermenu1 = menu1.CreateIterator();
            PrintMenu(intermenu1);
            UserIterator intermenu2 = menu2.CreateIterator();
            PrintMenu(intermenu2);
            sw.Stop();

            Console.WriteLine(sw.Elapsed.ToString());
        }

        public void PrintMenu2()
        {
            System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
            sw.Start();
            foreach (Item i in menu1.GetItems())
            {
                PrintMenu(i);
            }

            foreach (Item i in menu2.GetItems())
            {
                PrintMenu(i);
            }
            sw.Stop();

            Console.WriteLine(sw.Elapsed.ToString());
        }

        private void PrintMenu(Item item)
        {
            Console.WriteLine(item.GetName());
            Console.WriteLine(item.GetPrice());
        }

        private void PrintMenu(UserIterator inter)
        {
            while (inter.HasNext())
            {
                Item item = (Item)inter.Next();
                Console.WriteLine(item.GetName());
                Console.WriteLine(item.GetPrice());
            }
        }
    }
}

File: week9/younggeun0/iterator/Waitress.java
package week9.younggeun0.iterator;

public class Waitress {
	PancakeHouseMenu pancakeHouseMenu;
	DinerMenu dinerMenu;
	
	public Waitress(PancakeHouseMenu pancakeHouseMenu, DinerMenu dinerMenu) {
		this.pancakeHouseMenu = pancakeHouseMenu;
		this.dinerMenu = dinerMenu;
	}
	
	public void printMenu() {
		Iterator pancakeIterator = pancakeHouseMenu.createIterator();
		Iterator dinerIterator = dinerMenu.createIterator();
		System.out.println("메뉴\n---\n아침 메뉴");
		
		// 각 반복자를 가지고 오버로드된 printMenu 메소드를 호출 
		printMenu(pancakeIterator);
		System.out.println("\n점심 메뉴");
		printMenu(dinerIterator);
	}
	
	public void printMenu(Iterator iterator) {
		while (iterator.hasNext()) { // 이제 반복문이 한번만 있으면 됨
			MenuItem menuItem = (MenuItem)iterator.next();
			System.out.print(menuItem.getName() + ", ");
			System.out.print(menuItem.getPrice() + " -- ");
			System.out.println(menuItem.getDescription());
		}
	}
}
